src/dwarflines.cpp

```diff
@@ -99,41 +99,29 @@ bool _flushDWARFLines(const PEImage& img, mspdb::Mod* mod, DWARF_LineState& stat
 	for(size_t ln = 0; ln < state.lineInfo.size(); ln++)
 		printf("  %08x: %4d\n", state.lineInfo[ln].offset + 0x401000, state.lineInfo[ln].line);
 #endif
+	
+	int rc = 1;
+	unsigned int low_offset = state.lineInfo[0].offset;
+	unsigned short low_line = state.lineInfo[0].line;
 
-    int rc = 1;
-	unsigned int firstLine = state.lineInfo[0].line;
-	unsigned int firstAddr = state.lineInfo[0].offset;
-	unsigned int firstEntry = 0;
-	unsigned int entry = 0;
-	for(size_t ln = firstEntry; ln < state.lineInfo.size(); ln++)
+	for (size_t ln = 0; ln < state.lineInfo.size(); ++ln)
 	{
-		if(state.lineInfo[ln].line < firstLine || state.lineInfo[ln].offset < firstAddr)
-		{
-			if(ln > firstEntry)
-			{
-				unsigned int length = state.lineInfo[entry-1].offset + 1; // firstAddr has been subtracted before
-				if(dump)
-					printf("AddLines(%08x+%04x, Line=%4d+%3d, %s)\n", firstAddr, length, firstLine, entry - firstEntry, fname.c_str());
-				rc = mod->AddLines(fname.c_str(), segIndex + 1, firstAddr, length, firstAddr, firstLine,
-									(unsigned char*) &state.lineInfo[firstEntry],
-									(ln - firstEntry) * sizeof(state.lineInfo[0]));
-				firstLine = state.lineInfo[ln].line;
-				firstAddr = state.lineInfo[ln].offset;
-				firstEntry = entry;
-			}
-		}
-		else if(ln > firstEntry && state.lineInfo[ln].offset == state.lineInfo[ln-1].offset)
-			continue; // skip entries without offset change
-		state.lineInfo[entry].line = state.lineInfo[ln].line - firstLine;
-		state.lineInfo[entry].offset = state.lineInfo[ln].offset - firstAddr;
-		entry++;
+		auto& line_entry = state.lineInfo[ln];
+		line_entry.line -= low_line;
+		line_entry.offset -= low_offset;
 	}
-	unsigned int length = eaddr - firstAddr;
-	if(dump)
-		printf("AddLines(%08x+%04x, Line=%4d+%3d, %s)\n", firstAddr, length, firstLine, entry - firstEntry, fname.c_str());
-	rc = mod->AddLines(fname.c_str(), segIndex + 1, firstAddr, length, firstAddr, firstLine,
-					    (unsigned char*) &state.lineInfo[firstEntry],
-					    (entry - firstEntry) * sizeof(state.lineInfo[0]));
+
+	unsigned int high_offset = state.address - state.seg_offset;

```

Comment:
When flushing an intermediate chunk of line infos, isn't `state.address` from the first entry of the next chunk? That might produce invalid `address_range_length`.

Reply:
If by invalid you mean it underflows, then yes, `high_offset` can be equal to `low_offset` here so decrementing it and subtracting will produce `UINT_MAX`.  This is fine because the implementation immediately increments the length, bringing it back to 0.  I originally handled this case specially and left `address_range_length` at 0, but that added a bunch of length 1 ranges to the line info, which prevented the later longer line ranges from showing up (e.g. when using the ranges [0x100, 0x100] and [0x100, 0x110] we would only get line numbers for the instruction at 0x100, but the ranges [0x100, 0xff], [0x100, 0x110] give us line numbers for all instructions in the range [0x100, 0x110] and both line numbers would show up at 0x100).  It's weird and seems wrong, but it produces better debug info.

If by invalid you mean it crosses instructions not part of that line then no, there would be a `DW_LNE_end_sequence` before the new address, which would do a flush.

Reply:
I meant the case where `_flushDWARFLines` is called due to `entry.offset < first_entry.offset` which would cause `high_offset` to be smaller than `low_offset`. This is probably a rare case more likely with optimized code.
If it is expected and it's ok to have very large offsets in that case, too, a comment in the code would be nice.

Reply:
Good catch.  I've made it so decreasing addresses mark the previous address as the end and start a new sequence.